simple/stats/runner.py

```diff
@@ -261,100 +270,111 @@ def _vertical_specs_dcids(self,
     generate_svg_cache(self.db, self.svg_specialized_names)
 
-  # If the fh is a "special" file, append it to the self.special_handlers dict.
-  # Returns true if it is, otherwise false.
-  def _maybe_set_special_fh(self, fh: FileHandler) -> bool:
-    file_name = fh.basename()
-    file_type = self.special_file_to_type.get(file_name)
-    if file_type:
-      self.special_handlers[file_type] = fh
-      return True
-    return False
+  def _check_if_special_file(self, file: File) -> None:
+    for file_type in self.special_file_names_by_type.keys():
+      if file_type in self.special_files:
+        # Already found this special file.
+        continue
+      file_name = self.special_file_names_by_type[file_type]
+      if match(file, file_name):
+        self.special_files[file_type] = file
 
   def _run_all_data_imports(self):
-    input_fhs: list[FileHandler] = []
-    input_mcf_fhs: list[FileHandler] = []
-    for input_handler in self.input_handlers:
-      if not input_handler.isdir:
-        if self._maybe_set_special_fh(input_handler):
-          continue
-        input_file_name = input_handler.basename()
-        if input_file_name.endswith(".mcf"):
-          input_mcf_fhs.append(input_handler)
-        else:
-          input_fhs.append(input_handler)
+    input_files: list[File] = []
+    input_csv_files: list[File] = []
+    input_mcf_files: list[File] = []
+
+    for input_store in self.input_stores:
+      if input_store.isdir():
+        input_files.extend(input_store.as_dir().all_files(
+            self.include_input_subdirs))
       else:
-        for input_file in sorted(input_handler.list_files(extension=".csv")):
-          fh = input_handler.make_file(input_file)
-          if not self._maybe_set_special_fh(fh):
-            input_fhs.append(fh)
-        for input_file in sorted(input_handler.list_files(extension=".mcf")):
-          fh = input_handler.make_file(input_file)
-          if not self._maybe_set_special_fh(fh):
-            input_mcf_fhs.append(fh)
-        for input_file in sorted(input_handler.list_files(extension=".json")):
-          fh = input_handler.make_file(input_file)
-          self._maybe_set_special_fh(fh)
-
-      self.reporter.report_started(import_files=list(
-          map(lambda fh: fh.basename(), input_fhs + input_mcf_fhs)))
-      for input_fh in input_fhs:
-        self._run_single_import(input_fh)
-      for input_mcf_fh in input_mcf_fhs:
-        self._run_single_mcf_import(input_mcf_fh)
-
-  def _run_single_import(self, input_fh: FileHandler):
-    logging.info("Importing file: %s", input_fh.basename())
-    self._create_importer(input_fh).do_import()
-
-  def _run_single_mcf_import(self, input_mcf_fh: FileHandler):
-    logging.info("Importing MCF file: %s", input_mcf_fh.basename())
-    self._create_mcf_importer(input_mcf_fh, self.output_dir_fh,
+        input_files.append(input_store.as_file())
+
+    for input_file in input_files:
+      self._check_if_special_file(input_file)
```

Comment:
You probably want _check_if_special_file to return a bool and not do the other matches if it is a special file. Otherwise, if the special file is a csv or a mcf file (not the case today but future proofing), then it could be problematic.